queasars/job_shop_scheduling/problem_instances.py

```diff
@@ -0,0 +1,375 @@
+        return f"{str(self.operation)} starts at: {self.start_time} and ends at: {self.end_time}"
+
+
+class JobShopSchedulingResult:
+    """
+    Class representing an attempted solution to the given job shop scheduling problem instance
+
+    :param problem_instance: for which this schedule represents an attempted solution
+    :type problem_instance: JobShopSchedulingProblemInstance
+    :param schedule: which attempts to solve the given problem instance
+    :type schedule: dict[Job, tuple[ScheduledOperation, ...]]
+    """
+
+    def __init__(
+        self, problem_instance: JobShopSchedulingProblemInstance, schedule: dict[Job, tuple[ScheduledOperation, ...]]
+    ):
+
+        if set(problem_instance.jobs) != set(schedule.keys()):
+            raise JobShopSchedulingProblemException(
+                "The JobShopSchedulingResult must contain the same Jobs "
+                + "as the problem instance which it is a solution to!"
+            )
+
+        for job in problem_instance.jobs:
+            if job.operations != tuple(map(lambda x: x.operation, schedule[job])):
+                raise JobShopSchedulingProblemException(
+                    "The schedule for a Job must contain the same operations as the Job itself!"
+                )
+
+        self._problem_instance: JobShopSchedulingProblemInstance = problem_instance
+        self._schedule: dict[Job, tuple[ScheduledOperation, ...]] = schedule
+        self._is_valid: Optional[bool] = None
+        self._makespan: Optional[int] = None
+
+    @property
+    def problem_instance(self) -> JobShopSchedulingProblemInstance:
+        """
+        :return: the problem instance for which this result represents an attempted solution
+        :rtype: JobShopSchedulingProblemInstance
+        """
+        return self._problem_instance
+
+    @property
+    def schedule(self) -> dict[Job, tuple[ScheduledOperation, ...]]:
+        """
+        :return: schedule which attempts to solve the given problem instance
+        :rtype: dict[Job, tuple[ScheduledOperation, ...]]
+        """
+        return self._schedule
+
+    @property
+    def is_valid(self) -> bool:
+        """
+        :return: true, if the result adheres to all constraints of the JSSP and false otherwise
+        :rtype: bool
+        """
+        if self._is_valid is not None:
+            return self._is_valid
+        is_valid = self._is_valid_solution()
+        self._is_valid = is_valid
+        return is_valid
+
+    @property
+    def makespan(self) -> Optional[int]:
+        """
+        :return: the integer makespan if the result is valid and None otherwise
+        :rtype: Optional[int]
+        """
+        if not self.is_valid:
+            return None
+        if self._makespan is not None:
+            return self._makespan
+        makespan: int = max((scheduled_operations[-1].end_time for scheduled_operations in self.schedule.values()))
```

Comment:
This assumes that the `ScheduledOperation` in `self.schedule` are ordered. Can you document that assumption? Otherwise one may also introduce an additional class which makes sure that the access is always ordered (or use a type where the values are ordered or order the values within the constructor).

Reply:
There is an inherent order, the schedule has to adhere to. Namely the order of the operations within the job, which cannot be permutated. In the constructor i check that the tuples of ScheduledOperations adhere to that order (with respect to their index in the tuple, not the scheduled times!). Any schedule in which the scheduled times do not align with the order of the tuples is invalid. This is checked before calculating the makespan, which is why you can get the makespan as the maximum of the end times of the last operations over all Jobs.

Reply:
But you are correct, that this can be documented better. I will wait to change this until the name / necessity of the ScheduledOperation has been fully discussed, as this might still further change the documentation.

Reply:
I added according documentation to the class docstring